### PR TITLE
fix(developer): improve support for inferred key cap text

### LINKED ISSUE #7616: bug(developer): invalid key code in touch editor can generate `TypeError: code.charCodeAt is not a function`
Sentry Issue: [KEYMAN-DEVELOPER-QC](https://sentry.io/organizations/keyman/issues/3686705366/?referrer=github_integration)

```
TypeError: code.charCodeAt is not a function
  at Object.prepareKey (/app/source/toucheditor/lib/builder.js:1597:65)
  at Object.selectKey (/app/source/toucheditor/lib/builder.js:1390:12)
  at HTMLDivElement.<anonymous> (/app/source/toucheditor/lib/builder.js:1214:17)
  at HTMLDivElement.dispatch (/app/source/toucheditor/lib/jquery-1.10.2.js:5095:9)
  at HTMLDivElement.elemData.handle (/app/source/toucheditor/lib/jquery-1.10.2.js:4766:28)
```

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #7616.

Fixes a crash when attempting to infer key cap text from key id, and cleans up the support for inferred key text so that it works correctly in the designer.

When using `U_xxxx` style key ids, and where the key text is empty, the key cap should dynamically reflect the Unicode value of the key id. This has always been supported in KeymanWeb, but the implementation in the editor was somewhat unstable in the past.

Also adds a safeguard to KMW to stop invalid key ids causing a crash.

# User Testing

* **TEST_INFERRED_ID:** In touch layout editor, select a key, delete the key text, and set the key ID to `U_1234` or similar. The key should show the new character on the key cap, but the key text field should remain empty. Switch layers and ensure that the key text field remains empty for that key.

* **TEST_INFERRED_SUBKEY_ID:** In touch layout editor, select a key, add a subkey, delete the subkey text, and set the subkey ID to `U_1234` or similar. The subkey should show the new character on the subkey cap, but the subkey text field should remain empty. Switch layers and ensure that the subkey text field remains empty for that subkey.

### CODE REVIEW

**Review by darcywong00 — APPROVED** [review]
lgtm

File: web/source/osk/oskKey.ts [hunk 1]
```diff
@@ -411,7 +411,7 @@ namespace com.keyman.osk {
       const codePoints = id.substr(2).split('_');
       for(let codePoint of codePoints) {
         const codePointValue = parseInt(codePoint, 16);
-        if (((0x0 <= codePointValue) && (codePointValue <= 0x1F)) || ((0x80 <= codePointValue) && (codePointValue <= 0x9F))) {
+        if (((0x0 <= codePointValue) && (codePointValue <= 0x1F)) || ((0x80 <= codePointValue) && (codePointValue <= 0x9F)) || isNaN(codePointValue)) {
```
Comment (jahorton) [inline web/source/osk/oskKey.ts:414]: :+1:

**Review by jahorton — APPROVED** [review]

**Comment by bharanidharanj** [comment]
* **TEST_INFERRED_ID (FAILED):** Tested this PR with Keyman Developer 16.0.96-beta-test-7617 build in Windows 10 OS and here is my observation: 1. After setting the key ID to U_1234, the new character appeared on the key cap and the key text field remains empty. But, If I did the switching between different layers like shift or numeric the text field shows a letter of the shift layer text on the text field. 

   

* **TEST_INFERRED_SUBKEY_ID (PASSED):** Tested this as per the instructions and the subkey text field remains empty for that subkey after switching to different layers then return.

**Comment by mcdurdin** [comment]
> * **TEST_INFERRED_ID (FAILED):** Tested this PR with Keyman Developer 16.0.96-beta-test-7617 build in Windows 10 OS and here is my observation: 1. After setting the key ID to U_1234, the new character appeared on the key cap and the key text field remains empty. But, If I did the switching between different layers like shift or numeric the text field shows a letter of the shift layer text on the text field.

@bharanidharanj I'm unable to reproduce this problem as you describe. Can you do a video capture of the problem perhaps?

@keymanapp-test-bot retest TEST_INFERRED_ID

**Comment by bharanidharanj** [comment]
* **TEST_INFERRED_ID (FAILED):** Retested as per Marc's suggestion and I am able to reproduce the issue. I have attached the Video file (InferredKey.MP4) for reference.  Please, correct me if it is a wrong test. :-)

https://user-images.githubusercontent.com/19683143/200495946-7034eea6-9b71-43b9-b60d-28ac9e1ea901.mp4

**Comment by mcdurdin** [comment]
@bharanidharanj thank you for the video. Looking at the `U_1234` key that you changed (originally the `K_A` key), you never reselected it after switching back to the default layer, so we can't see if it has text stored in the "text" field, but I re-tested in the same way here and it was correct after switching back to the default layer. I'm going to go ahead and mark this as passed, but if there is still something wrong, please do open an issue :grin:

TEST_INFERRED_ID PASS

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.98-beta](https://keyman.com/downloads/releases/beta/16.0.98)

**Pull request merged**